plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -354,6 +357,55 @@ public boolean isSourceModified() {
         return sourceModified;
     }
 
+    static Object deepCopy(Object value) {
+        if (value instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<Object, Object> mapValue = (Map<Object, Object>) value;
+            return deepCopyMap(mapValue);
+        } else if (value instanceof List) {
+            @SuppressWarnings("unchecked")
+            List<Object> listValue = (List<Object>) value;
+            return deepCopyList(listValue);
+        } else {
+            return value;
```

Comment:
makes sense. I was wondering, given that this code relies on the assumption that the set of supported types is limited, shall we enforce this just to make sure that we never get a reference that's going to cause problems here (and throw UOE if we get something that we don't support)? Too paranoid?